labrad/node/__init__.py

```diff
@@ -152,168 +157,162 @@ def timeout(self):
-        # spawnProcess say that the full executable path is required, but
-        # this is not in fact the case. See:
-        # http://twistedmatrix.com/documents/current/api/twisted.internet.interfaces.IReactorProcess.spawnProcess.html
+        Note that the twisted docs for spawnProcess say that the full executable
+        path is required, but this is not in fact the case. See:
+
+            http://twistedmatrix.com/documents/current/api/twisted.internet.interfaces.IReactorProcess.spawnProcess.html
+        """
         if platformType == 'posix':
-            executable = self.args[0]
+            return self.args[0]
         elif platformType == 'win32':
-            executable = None
+            return None
         else:
             raise RuntimeError("Unsupported platform %s" % platformType)
 
-        self.proc = reactor.spawnProcess(self, executable, self.args,
+    def set_status(self, value):
+        """Transition to the given status and notify message listeners."""
+        message = self.STATUS_MESSAGES[value]
+        self.status = value
+        self.logger.info(value)
+        self.on_message(self, message)
+
+    @inlineCallbacks
+    def start(self):
+        self.set_status('STARTING')
+        self.logger.info("path: {}".format(self.path))
+        self.logger.info("args: {}".format(self.args))
+
+        # start listening for server connect messages
+        msg_ctx = self.client.context()
+        msg_id = msg_ctx[1]
+
+        connected = defer.Deferred()
+
+        def on_server_connect(message_ctx, msg):
+            _ID, name = msg
+            if name == self.name:
+                connected.callback(None)
+
+        message = 'Server Connect'
+        manager = self.client.manager
+        manager.addListener(on_server_connect, context=msg_ctx)
+        yield manager.subscribe_to_named_message(
+            message, msg_id, True, context=msg_ctx)
+
+        # start the server process
+        self.proc = reactor.spawnProcess(self, self.executable, self.args,
                                          env=self.full_env, path=self.path)
-        timeoutCall = reactor.callLater(self.timeout, self.kill)
+
+        # wait for the server to connect, shutdown, or timeout
+        selected = yield mux.select(
+            connected=connected,
+            shutdown=self.on_shutdown(),
+            timeout=self.timeout)
+
+        # stop listening for server connect messages
         try:
-            yield self.startup
-            self.emitMessage('server_started')
-        except:
-            self.emitMessage('server_stopped')
-            raise
-        finally:
-            self.starting = False
-            if timeoutCall.active():
-                timeoutCall.cancel()
+            manager.removeListener(on_server_connect, context=msg_ctx)
+            yield manager.subscribe_to_named_message(
+                message, msg_id, False, context=msg_ctx)
+        except Exception as e:
```

Comment:
Good catch; I used it in a print statement previously, but not needed with the logging statement. Removed.